xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp

```diff
@@ -217,19 +220,31 @@ float CEngineStats::GetCacheTotal()
 
 float CEngineStats::GetMaxDelay() const
 {
-  return static_cast<float>(MAX_CACHE_LEVEL) + static_cast<float>(MAX_WATER_LEVEL) +
-         m_sinkCacheTotal;
+  return MAX_CACHE_LEVEL + MAX_WATER_LEVEL + m_sinkCacheTotal;
 }
 
 float CEngineStats::GetWaterLevel()
 {
   std::unique_lock<CCriticalSection> lock(m_lock);
+
   if (m_pcmOutput)
+  {
     return static_cast<float>(m_bufferedSamples) / m_sinkSampleRate;
+  }
   else
-    return static_cast<float>(m_bufferedSamples *
-                              m_sinkFormat.m_streamInfo.GetDuration(m_sinkNeedIecPack)) /
-           1000;
+  {
+    float wl = static_cast<float>(m_bufferedSamples *
+                                  m_sinkFormat.m_streamInfo.GetDuration(m_sinkNeedIecPack)) /
+               1000;
+
+    if (m_sinkFormat.m_streamInfo.m_type == CAEStreamInfo::STREAM_TYPE_TRUEHD)
+    {
+      if (wl >= MAX_WATER_LEVEL)
+        wl = MAX_WATER_LEVEL - 0.00001f;
```

Comment:
I would add a sanity check as well to see if TrueHD actually overruns MAX_WATER_LEVEL by far more milliseconds.

Reply:
The more I check the code, the more I get the feeling that the waterlevel is not relevant for TrueHD. For me it seems that Paplayer is the only guy doing something special here with the levels. See also forum feedback without that change - I will look more in detail over the next few days. Cutting an "arbitrary" high value down to WATER_LEVEL is not good - it's another workaround for something we don't have yet fully understood.

Reply:
Already explained but anyway:

Shield has a very specific micro-stutter type _only_ with TrueHD.  (some people has also other types of stuttering due high CPU but are unrelated to this).

Micro-stutter of "TrueHD only" can be fixed with "ratios" in AudioTrack and half `GetDutation()`

Now something NEW has been discovered:

These micro-stutter of "TrueHD only" ALSO are gone reverting ratio and GetDuration() BUT changing water level behavior.

Seems you not like water level and propose revert also ratio and GetDuration().... 

Then inevitably you get the TrueHD micro-stutter again and Shield is not usable.

Reply:
>  See also forum feedback without that change

Because in forum people are using advancedsettings with value 90 -100 that only should be used to fix the others type of video stuttering (due high CPU usage). Also this setting introduces a/v sync issues of same magnitude that value of setting, so is not usable for everyone.

Reply:
from forum:
> That build only changes one thing to yesterday: It uses 24 audio units for TrueHD again

This is not usable because re-introduce random audio dropouts due MAT packer latency:

24 + 24 ----->  MAT packer ------> 24

there are 24 units of latency 20ms vs 10ms using  12 + 12

Reply:
I agree for the last comment: 12 Units it needs to be. Forum has shown that.
For the first in correlation with the discon - I don't. This is needed for them no matter if ratio is applied or not. It only caused in my POV from the large SINK error that is caused by nvidia's AudioTrack implementation: 1,2,1,80 ms blocking time.

Reply:
> This is needed for them no matter if ratio is applied or not.

Right but this is separated issue (not related with TrueHD) and only some users are affected by this.

In my setup I have micro-stutter free playback with default 10ms ONLY if water level modification is applied.

Reply:
**option A:** water level workaround and ratio / half duration reversion (this PR)

**option B:** close PR and remains ratio / half duration as current master & Nexus

**option C:** no water level workaround and ratio / half duration reversion. => **NOT POSSIBLE**

**option D:** use 24 audio units in MAT packer => **NOT POSSIBLE**

**option E:** fix true root cause of PAPlayer progress indicator and "a/v sync oscillation" and debug info OSD oscillation. ??

Note that same thing happens in PAPlayer happens also in VideoPlayer, only difference is that is not visible because progress indicator is based on video time stamps (not audio). Then is "not visible" but same issue exists.

Reply:
I want to look into the DVD Passthrough Codec this evening. Here the PTS timestamps are "computed" ... hopefully they don't role forth and back ...

If you check where WaterLevel is basically used in AE, the you see that passthrough is the "else" case ... where it should not play a role. In PaPlayer the audio does not stutter, but we see that the playbar is jumpy - as PaPlayer updates the values by itself (setting his status) most likely we can find out, why the value flactuates.

To see if E is really an option.

For the ratio workaround: ALSA and Windows need adjusting ... as well, this is currently not correct then, only AudioTrack works around this. As show on the forum: Impact of 100 ms further delay, cause AE does not know anymore, that Sink has 100 ms audio more than it reports ... this is why I don't like the ratio workaround.

Reply:
For PAPlayer:
```
      si->m_framesSent += si->m_audioFormat.m_streamInfo.GetDuration() / 1000 *
                          si->m_audioFormat.m_streamInfo.m_sampleRate;
```
```
free_space = (double)si->m_stream->GetSpace() *
                     si->m_audioFormat.m_streamInfo.GetDuration() / 1000;
```

The first one is always: 0.02 * 192 khz - so both assume again that every frame contains 20 ms of data.
The second one is always: GetSpace x 0.02 - but there is more: the first method assume a direct relation ship between "payload time" and samplerate, which is only the cause if the payload of a frame matches the duration of the data. The second method instead only again is intersted in the number of free sample by duration.

While GetSpace is from the Stream itself:
```
unsigned int CActiveAEStream::GetSpace()
{
  std::unique_lock<CCriticalSection> lock(m_streamLock);
  if (m_format.m_dataFormat == AE_FMT_RAW)
    return m_streamFreeBuffers;
  else
    return m_streamFreeBuffers * m_streamSpace;
}

The m_streamFreeBuffers are incremented whenever possible. AE is blocking below to fillup the stream again. until the cache has MAX_CACHE_LEVEL is reached. So being larger than CACHE_LEVEL is a good thing.
```
    // provide buffers to stream
    float time = m_stats.GetCacheTime((*it));
    CSampleBuffer *buffer;
    if (!(*it)->m_drain)
    {
      float buftime = (float)(*it)->m_inputBuffers->m_format.m_frames / (*it)->m_inputBuffers->m_format.m_sampleRate;
      if ((*it)->m_inputBuffers->m_format.m_dataFormat == AE_FMT_RAW)
        buftime = (*it)->m_inputBuffers->m_format.m_streamInfo.GetDuration() / 1000;
      while ((time < static_cast<float>(MAX_CACHE_LEVEL) || (*it)->m_streamIsBuffering) &&
             !(*it)->m_inputBuffers->m_freeSamples.empty())
      {
        buffer = (*it)->m_inputBuffers->GetFreeBuffer();
        (*it)->m_processingSamples.push_back(buffer);
        (*it)->m_streamPort->SendInMessage(CActiveAEDataProtocol::STREAMBUFFER, &buffer, sizeof(CSampleBuffer*));
        (*it)->IncFreeBuffers();
        time += buftime;
      }
    }
```
And with this very same assumption paplayer clustters this in its time info, here:
```
  double time = ((double)m_currentStream->m_framesSent / (double)m_currentStream->m_audioFormat.m_sampleRate);
  if (m_currentStream->m_stream)
    time -= m_currentStream->m_stream->GetDelay();
  time = time * 1000.0;

  m_playerGUIData.m_time = (int64_t)time; //update for GUI
 ```

Reply:
And bye -> lunchbreak is over :-)

Reply:
Other important point:

PAPlayer jumps in progress issue exists from before TrueHD MAT packer first rework in v19.3. See: https://github.com/xbmc/xbmc/issues/20391 Then is unrelated to use 12 or 24 audio units and even is unrelated to use IEC or RAW.

> Kodi version:
> 19.0, 19.1, 19.2
> 18.4, 18.5 etc

This means that what we are trying to patch here is a bug in AE core that has always been latent.

Tomorrow try to split water level commit in two:

first one to fix float "rounding" issues, that is

`if (float A == float B) ` is incorrect by design

`if (abs(A - B) < epsilon)`    <------ should be

in a second commit specific behavior changes for TrueHD only (this can be reverted when a better solution is found)

Reply:
I think we can ommit the second commit for now and just let it jump - and then properly fixing paplayer. From my investigation the problem is with framesent using samplerate for conversion and freespace the duration ... so one is tied to the size of the package that other not. This can't work.

AE does not check == but use while (time < MAXLEVEL) which is perfectly fine, as this will end when it is <= MAXLEVEL.

Reply:
Sorry, but you are not understanding the problem. PAPlayer doesn't matter to me. What I want to fix are the problems in VideoPlayer with TrueHD

Reply:
Have a look here: 09b0276868c413a168c9cbc640c878d737fe1836

Reply:
Same BUG causes:

- PAPlayer jumps.
- Debug info OSD a/v sync fluctuates much with TrueHD  (playing video).
- Android (Shield) micro-stutter playing video + TrueHD.

ALL THIS is due water level issue

Reply:
Please see the second part of the commit I just linked. Here the TrueHD special case needs to go in again and only half the duration needs to be added when the 12 frame version is running.

Reply:
The original code added the timing by 1/24 ... but then it was changed to only once per 61440 add one frame. Now we come twice - therefore adding half of it - should fix the timing again.

Reply:
```
      bool rawPktComplete = false;
      {
        std::unique_lock<CCriticalSection> lock(m_statsLock);
        if (m_format.m_dataFormat != AE_FMT_RAW)
        {
          m_currentBuffer->pkt->nb_samples += minFrames;
          m_bufferedTime +=
              static_cast<float>(minFrames) / m_currentBuffer->pkt->config.sample_rate;
        }
        else
        {
          if (m_format.m_streamInfo.m_type == CAEStreamInfo::STREAM_TYPE_TRUEHD && isIEC)
          {
            m_currentBuffer->pkt->nb_samples += 30720;
            m_bufferedTime += m_format.m_streamInfo.GetDuration() / 1000 / 2;
            if (m_currentBuffer->pkt->nb_samples / 2560 == 24)
              rawPktComplete = true;
          }
          else
          {
            m_bufferedTime += static_cast<float>(m_format.m_streamInfo.GetDuration()) / 1000;
            m_currentBuffer->pkt->nb_samples += minFrames;
            rawPktComplete = true;
          }
        }
      }
 ```
 something like this, that way the m_bufferedTime will only increase half and the rest of the condition can stay as is ...

Reply:
This looks promising 👍   I'll try it when I have time

Thank you!

Reply:
Nothing to thank - we have the same goal :-)

Reply:
Unfortunately it doesn't work, some reasons I understand others I don't:

`rawPktComplete = true;` needs in all half packets because in

https://github.com/xbmc/xbmc/blob/fa592d344eb0b5e2bc8864cb304173e696e36a79/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESink.cpp#L951

needs to processed every half packet (12 units). Then `ActiveAESink` half packet is considered as `rawPktComplete`

`m_format.m_streamInfo.GetDuration() / 1000 / 2;`  seems very logic and correct but causes video player hangs (not starts playback any video frame) / black screen.

PAPlayer works but not resolves main issues: progress indicator stills jumps and duration is wrong (speed x2). Audio is OK.

Reply:
See comment in a new commit https://github.com/xbmc/xbmc/pull/22705/commits/dbbb535d272f20d652b37cc3bb5d55fc4890c1c4 to understand why jumps progress indicator:

jumps is because a/v sync in NOT calculated all the time because condition 

`m_stats.GetWaterLevel() < MAX_WATER_LEVEL  `

is not true all the time.

Reply:
My understanding is that with TrueHD the water level constantly fluctuates around the `MAX_WATER_LEVEL`: a little above and a little below but that's okay. The only problem is that AE was not designed with this in mind.

Reply:
> Unfortunately it doesn't work, some reasons I understand others I don't:
> 
> `rawPktComplete = true;` needs in all half packets because in
> 
> https://github.com/xbmc/xbmc/blob/fa592d344eb0b5e2bc8864cb304173e696e36a79/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESink.cpp#L951
> 
> needs to processed every half packet (12 units). Then `ActiveAESink` half packet is considered as `rawPktComplete`
> 
> I mean that `ActiveAESink` only is called half of times and audio is corrupt.
> 
> `m_format.m_streamInfo.GetDuration() / 1000 / 2;` seems very logic and correct but causes VideoPlayer hangs (not starts playback any video frame) / black screen.
> 
> PAPlayer works but not resolves main issues: progress indicator stills jumps and duration is wrong (speed x2). Audio is OK.

Okay - that actually makes sense. We pull out 12 units all the time, but our IEC packer, the final guy in the chain at the end only outputs something after 24 packets were added. So therefore faking "Complete" to VideoPlayer all the time could be done, but then only adding half of the time?

Like:
```
      {
        std::unique_lock<CCriticalSection> lock(m_statsLock);
        if (m_format.m_streamInfo.m_type == CAEStreamInfo::STREAM_TYPE_TRUEHD && isIEC)
        {
          m_currentBuffer->pkt->nb_samples += minFrames;
          m_bufferedTime += m_format.m_streamInfo.GetDuration() / 1000 / 2;
          rawPktComplete = true;
        }
        else
        {
        .... // current code
        }
      }
 ```

That should achieve that rawPktComplete as always true to enter next stage, but not adding a wrong delay?

Reply:
As commented 

`m_bufferedTime += m_format.m_streamInfo.GetDuration() / 1000 / 2;`

causes VideoPlayer hangs / black screen

Reply:
No hurry. Then on the other side of things, when running through "the stages" someone still has the wrong information ... will also check again on the weekend.

Reply:
Code to clamp water level to MAX_WATER_LEVEL has been removed, then this issue is closed.

@fritsch, please review as new PR.  

Open your mind and do not bring preconceived ideas 😊